documentation/docs/extensibility.md

```diff
@@ -1,86 +1,199 @@
+* For SAP Cloud SDK Pipeline, you can find them in [this GitHub search query](https://github.com/SAP/cloud-s4-sdk-pipeline-lib/search?q=%22def+stageName+%3D%22).
 
-      //execute original stage as defined in the template
-      params.originalStage()
+The centrally maintained pipeline checks if such a file exists and executes it, if present.
+A parameter of type `Map` that contains the following keys is passed to the extension:
 
-      //access overall pipeline script object
-      echo "Branch: ${params.script.commonPipelineEnvironment.gitBranch}"
+* `script`: defines the global script environment of the `Jenkinsfile` run. This makes sure that the correct configuration environment can be passed to project "Piper" steps and allows access to for example the `commonPipelineEnvironment`.
+* `originalStage`: this will allow you to execute the "original" stage at any place in your script. If omitting a call to `originalStage()` only your code will be executed instead.
+* `stageName`: name of the current stage
+* `config`: configuration of the stage and general config (including all defaults)
 
-      echo "End - Extension for stage: ${params.stageName}"
-  }
-  return this
-  ```
+Here is a simple example for such an extension, which you can use as a starting point:
+
+```groovy
+void call(Map params) {
+  //access stage name
+  echo "Start - Extension for stage: ${params.stageName}"
+
+  //access config
+  echo "Current stage config: ${params.config}"
+
+  //execute original stage as defined in the template
+  params.originalStage()
+
+  //access overall pipeline script object
+  echo "Branch: ${params.script.commonPipelineEnvironment.gitBranch}"
+
+  echo "End - Extension for stage: ${params.stageName}"
+}
+return this
+```
+
+!!! note "`return this`"
+    Don't forget the `return this` which is required at the end of _all_ extension scripts.
+    This is due to how Groovy loads scripts internally.
 
 !!! note "Init stage cannot be extended"
-    Please note, the `Init` stage among other things also checks out your current repository.<br />Thus it is not possible to use extensions on this stage.
+    Please note, the `Init` stage among other things also checks out your current repository including your extensions.<br />
+    Thus it is not possible to use extensions on this stage.
+
+### Practical example
+
+For a more practical example, you can use extensions in SAP Cloud SDK Pipeline to add custom linters to the pipeline.
+
+A linter is a tool that can check the source code for certain stylistic criteria, and many teams chose to use a linter to ensure a common programming style.
+
+As an example, if you want to use [Checkstyle](https://checkstyle.sourceforge.io/) in your codebase, you might use an extension similar to this one in a file called `.pipeline/extensions/lint.groovy` in your project:
+
+```groovy
+def call(Map parameters) {
+
+    parameters.originalStage() // Runs the built in linters
+
+    mavenExecute(
```

Comment:
```suggestion
    // Then run Checkstyle linter
    mavenExecute(
```

Reply:
Does that comment explain anything? I think from the context it is clear that this is happening, isn't it?

Reply:
It is not clear to me, probably because I have very little experience with linters. What is confusing is that you call mavenExecute to use the Checkstyle linter. But, adding the explanatory text below the code solves this issue.

Reply:
Then maybe go all the way with the comment and hint at the goal, i.e. "Then run Checkstyle linter via mvn goal"